modules/sw360/sw360-client/src/main/java/org/eclipse/sw360/antenna/sw360/client/rest/SW360Client.java

```diff
@@ -0,0 +1,263 @@
+
+    /**
+     * Generates a URL pointing to a specific resource of the SW360 server.
+     * This method concatenates the given path segments and appends them to the
+     * base REST URL of the server. It can be used by derived classes to
+     * generate the URLs to be requested.
+     *
+     * @param paths the path segments to add to the base URL
+     * @return a string with the resulting URL
+     */
+    protected String resourceUrl(String... paths) {
+        return getClientConfig().getRestURL() + URL_PATH_SEPARATOR + String.join(URL_PATH_SEPARATOR, paths);
+    }
+
+    /**
+     * Implements the actual request execution logic including a retry
+     * mechanism if the current access token may have expired. This method
+     * generates a future that obtains the current access token and executes
+     * the request defined by the parameters with it. If the {@code canRetry}
+     * parameter is <strong>true</strong> (which it is only for the first time
+     * the request is executed), the future is enhanced with special retry
+     * logic. This logic checks whether a failure has occurred indicating an
+     * expired access token. If this is the case, another request execution is
+     * chained to the original future.
+     *
+     * @param producer  the {@code RequestProducer}
+     * @param processor the {@code ResponseProcessor}
+     * @param tag       a tag to identify the request
+     * @param canRetry  a flag whether a retry is possible
+     * @param <T>       the type of the result
+     * @return a future with the result of the request
+     */
+    private <T> CompletableFuture<T> executeWithExpiredTokenCheck(RequestProducer producer,
+                                                                  ResponseProcessor<T> processor,
+                                                                  String tag,
+                                                                  boolean canRetry) {
+        LOG.debug("Executing request '{}'{}.", tag, canRetry ? "" : " (retry)");
+
+        CompletableFuture<T> futRequest = getTokenProvider().doWithToken(accessToken ->
+                getClientConfig().getHttpClient().execute(accessToken.tokenProducer(producer),
+                        applyChecks(processor, tag, accessToken)));
+        return canRetry ?
+                FutureUtils.withConditionalFallback(futRequest,
+                        this::checkIfRetry,
+                        () -> executeWithExpiredTokenCheck(producer, processor, tag, false)) :
+                futRequest;
+    }
+
+    /**
+     * Checks whether a request needs to be retried because the access token
+     * may have expired. The method checks whether the request failed with the
+     * special {@code RequestUnauthorizedException} exception.
+     *
+     * @param exception the exception the request failed with if any
+     * @return a flag whether a retry should be attempted
+     */
+    private boolean checkIfRetry(Throwable exception) {
```

Comment:
This is a heavy side effect, a check method that invalidates the token.

Reply:
Yup, and suggestions for improvements are welcome ;-)

Fact is that we need to manage a state - the current access token. At some point this state has to be updated.

Reply:
This has been reworked. Token invalidation is now handled by the AccessTokenProvider class.